zerver/lib/rate_limiter.py

```diff
@@ -160,6 +164,106 @@ def rate_limit_entity(cls, entity_key: str, rules: List[Tuple[int, int]],
+        if cls.last_gc_time <= now - time_window / max_count:
+            cls.last_gc_time = now
+            cls.reset_times = {key: time_dict for (key, time_dict) in cls.reset_times.items()
+                               if time_dict[(time_window, max_count)] > now}
+
+        if entity_key in cls.timestamps_blocked_until:
+            # Check whether the key is manually blocked.
+            if now < cls.timestamps_blocked_until[entity_key]:
+                blocking_ttl = cls.timestamps_blocked_until[entity_key] - now
+                return True, blocking_ttl
+            else:
+                # Remove from timestamps_blocked_until if the blocking
+                # time has been expired.
+                del cls.timestamps_blocked_until[entity_key]
+
+        if entity_key in cls.reset_times:
+            # Check whether the new request would overfill its bucket.
+            new_reset = max(cls.reset_times[entity_key][(time_window, max_count)],
+                            now) + time_window / max_count
+        else:
+            # The bucket was empty, so one request cannot overfill it.
+            new_reset = now + time_window / max_count
+
+        if new_reset > now + time_window:
+            # Compute for how long the bucket will remain filled.
+            time_till_free = new_reset - time_window - now
+            return True, time_till_free
+
+        cls.reset_times[entity_key] = {(time_window, max_count): new_reset}
+        return False, 0.0
+
+    @classmethod
+    def get_api_calls_left(cls, entity_key: str, range_seconds: int,
+                           max_calls: int) -> Tuple[int, float]:
+        now = time.time()
+        reset_time = cls.reset_times[entity_key][(range_seconds, max_calls)]
+        calls_remaining = (now + range_seconds - reset_time) * max_calls // range_seconds
+        return int(calls_remaining), reset_time
+
+    @classmethod
+    def block_access(cls, entity_key: str, seconds: int) -> None:
+        now = time.time()
+        cls.timestamps_blocked_until[entity_key] = now + seconds
+
+    @classmethod
+    def unblock_access(cls, entity_key: str) -> None:
+        del cls.timestamps_blocked_until[entity_key]
+
+    @classmethod
+    def clear_history(cls, entity_key: str) -> None:
+        now = time.time()
+
+        for time_window, max_calls in cls.reset_times[entity_key]:
+            cls.reset_times[entity_key][(time_window, max_calls)] = now - time_window
+
+    @classmethod
+    def rate_limit_entity(cls, entity_key: str, rules: List[Tuple[int, int]],
+                          max_api_calls: int, max_api_window: int) -> Tuple[bool, float]:
+
+        for time_window, max_count in rules:
+            ratelimited, time = cls.need_to_limit(entity_key, time_window, max_count)
+
+            if ratelimited:
+                statsd.incr("ratelimiter.limited.%s" % (entity_key,))
+                break
+
+        return ratelimited, time
```

Comment:
I'm not sure the `time` we will return is correct. This terminates on the first failed rule and returns the "time to freedom" for this rule. But the key may be at the limit of more than one rule, and possibly the "time to freedom" for one of the other rules could be longer than for the rule we terminated the iteration on. So I think we should iterate over all the rules, and `time` we return should be the `max` of the list of times for each rule?

Reply:
@mateuszmandera but the `RedisRateLimiterBackend` implementation doesn't returns the max `time_to_freedom` if more than one rule is violated. It just returns the first `time_to_freedom` value just like I did here. For example,
```
'api_by_user': [
        (60, 200),
        (30, 1),
        (100, 1)
    ],
```
For the rules above, `RedisRateLimiterBackend` will return `True, 30.0` instead of `True, 100.0`.

Reply:
I think it's ok if the current code only reports the time for the particular rule that's being violated.  Unless these changes are somehow making the situation worse, we can probably handle this with a follow-up commit.  But I'll defer to @mateuszmandera here, since I'm not super familiar with the code here.

Reply:
@showell @mateuszmandera I think we should handle it like `RedisRateLimiterBackend` does. It goes through the rules from shortest from longest. For example,
```
'api_by_user': [
        (60, 200),
        (30, 1),
        (100, 1)
    ],
```
For the rules above, `RedisRateLimiterBackend` will return `time_till_freedom` for the 2nd rule first. After 30 seconds are passed, it will return `time_till_freedom` for the 3rd rule.

Reply:
This is indeed how the redis code seems to do it, but I think it should be considered a bug. But I agree that we could keep this the way it is, both rate limiting implementations being consistent in this behavior, and fix both later in a follow-up commit.